### PR TITLE
Remove 0.4 Deprecations

### PR DESCRIPTION
**Pull request by jakebolewski**
Follow up on #13052.

Two tests are failing in arrayops.jl which check the return types of `setindex!` and `getindex`.  I commented them out for now.

~~Still need to remove syntax deprecations.~~~

### CODE REVIEW

**Comment by JeffBezanson** [comment]
b5a251bf1498dca907ecceb748eeb2015bca8c9a should fix the arrayops test.

**Comment by IainNZ** [comment]
@jakebolewski could you clarify: these are things that were deprecated between 0.3 release and 0.4 release, and will no longer work on 0.5?

**Comment by IainNZ** [comment]
Its the former, right? For the latter, we had https://github.com/JuliaLang/julia/pull/11046. That came mid-release-cycle, @tkelman wanted it at the end of the release cycle, and I guess this is at the start of the release cycle?

**Comment by JeffBezanson** [comment]
It's an interesting question. Technically, it's correct to remove 0.4 deprecations on master immediately, but in practice people might prefer a slower transition. I like to rip the bandaid right off, others like to do it slowly.

**Comment by IainNZ** [comment]
I'm in Camp Bandaid FWIW, and I think doing it now is fairly principled and helps get 0.5 out that much sooner. Perhaps has the social effect of keeping people well away from 0.5 too, because it'll knock out even more packages.

**Comment by mbauman** [comment]
At a minimum, it'd be nice to hold off on this until I (and/or others) get #13047 a little closer.  Those doc conversions can still get backported easily.

That said, some of the 0.5 changes I'd like to make will depend upon removing these deprecations.  Count me in for camp bandaid.

**Comment by ScottPJones** [comment]
Please, rip it off!  People have their 0.4 RC1 goodness already, bandaid off will make it easier to find stuff to fix in packages.

**Comment by jakebolewski** [comment]
Yes, my hidden intention is to force people into using the v0.4-rc right away for their primary work.

@mbauman I intend to merge the doc changes before this.

**Comment by JeffBezanson** [comment]
Oh, we also need to flip the `_oldstyle_array_vcat_` flag.

**Comment by sbromberger** [comment]
Just to clarify, do the devs WANT people to use 0.5, or would they (you) prefer us to stick with 0.4? After reading the comments here I'm confused as to how I should proceed (as a non-dev).

**Comment by johnmyleswhite** [comment]
Please do not start using 0.5 immediately.

**Comment by quinnj** [comment]
If you plan on helping with developing new 0.5 features (planned or un-planned), or are happy helping test new changes that get merged, then feel free, I say. I'd say the majority of "package" developers probably don't fall in this camp, however, and need/want something more predictable to build packages against.

**Comment by sbromberger** [comment]
As long as  https://github.com/JuliaLang/julia/pull/13036 gets backported soon, I'm happy to stick with 0.4. :)

**Comment by tkelman** [comment]
For crying out loud, we need to solve #11200 before doing this. Otherwise we're breaking code vindictively.

**Comment by jakebolewski** [comment]
Ok this is done.

Break what code? Standard library code?  I don't see why we have to have a solution ready before moving forward during the 0.5 release cycle.  Certainly before tagging a release.

**Comment by JeffBezanson** [comment]
I think we can come up with a way to deprecate bindings and backport it to 0.4, so that somewhere in the middle of the cycle 0.4 users will start seeing more warnings.

**Comment by tkelman** [comment]
No, I'm saying removing deprecations that never even gave a warning is breaking user code no matter when we do it. So we shouldn't rush this, since we have time to try to fix #11200 first and make this less painful. There's zero advantage to purposefully breaking packages and user code on 0.5 as early as possible. It'll make regression testing of codegen and package manager rewrites more annoying, as one example.

**Comment by jakebolewski** [comment]
We need to free up the syntax to do get other changes in the pipeline.  If we want to release by Feb then we realistically have a little over 4 months to make that happen.

All major packages that would be prudent to backtest on already depend on Compat.  The package manager rewrite (if it doesn't depend on deprecated 0.4 functionality) should just work or work with minor modifications.

**Comment by tkelman** [comment]
Has Compat been updated for all of these? I'm suggesting we do this systematically, if some deprecations are needed for high-priority array changes then do them first. Most of those should already be giving warnings so I'm not as opposed to removing them when needed. Removing absolutely all of the deprecations en masse, including ones that never gave warnings, just isn't needed right now.

**Comment by jakebolewski** [comment]
I guess my only concern is I don't see how we can meet the goals of the 0.5 release while maintaining the level of stability in the 0.4 release cycle.

File: test/spawn.jl [hunk 1]
```diff
@@ -208,37 +208,39 @@ close(f)
...
+    import Base.zzzInvalidIdentifier
+    """
+    try
+        (in,p) = open(`$exename --startup-file=no`, "w")
+        write(in,cmd)
+        close(in)
+        wait(p)
+    catch
+        error("IOStream redirect failed. Child stderr was \n$(readall(fname))\n")
+    finnaly
```
Comment (tkelman) [inline test/spawn.jl]: typo

**Comment by tkelman** [comment]
Another reason I dislike doing this right away is that I intend to use PackageEvaluator results on nightly as a rough filter for backporting changes from master to release-0.4. This gets less useful over time as the branches diverge, but I plan on writing up a formal backporting policy soonish that will include "don't backport code changes before PackageEvaluator has had a chance to run with them on nightly" as a guideline (mostly relevant after 0.4.0 is out when there's not as much of a rush, but still more info there than just CI even now). Breaking a bunch of code before it's actually necessary just adds a whole bunch of noise and makes it harder to judge whether or not other changes could cause any problems.

**Comment by phobon** [comment]
Here's a data point regarding #11200: ~350 packages still use the characters `"::String"` in their code, presumably because the developers have never seen a deprecation warning telling them to switch. I don't know how many of that number were fully vetted for 0.4, but the list includes the names `DataFrames`, `Escher`, `Gadfly`, `SFML`, and more.

**Comment by tkelman** [comment]
Mostly superseded by #16585 and #16605, though the parser changes here are still needed I think.

The command-line argument removals can't happen until after there's been a release where they print warnings, #16596.

**Pull request closed without merging**